File: leetcode/problems/problem-specific/15.3-sum/15.3-sum.go
package main

import "sort"

func threeSum(nums []int) [][]int {
	if len(nums) < 3 {
		return [][]int{}
	}

	res := [][]int{}
	sort.Ints(nums)
	for i := 0; i < len(nums)-1; i++ {
		left, right := i+1, len(nums)-1
		for left < right {
			totalSum := nums[i] + nums[left] + nums[right]
			if totalSum == 0 {
				triplet := []int{nums[i], nums[left], nums[right]}
				res = append(res, triplet)
				leftVal := nums[left]
				for left < len(nums) && nums[left] == leftVal {
					left++
				}
				rightVal := nums[right]
				for right > left && nums[right] == rightVal {
					right--
				}
			} else if totalSum < 0 {
				left++
			} else {
				right--
			}
		}

		for i+1 < len(nums)-1 && nums[i] == nums[i+1] {
			i++
		}
	}

	return res
}

File: leetcode/problems/language-specific/python/15.3-sum.py
#
# @lc app=leetcode id=15 lang=python3
#
# [15] 3Sum
#

# @lc code=start

# * Two Pointers Solution | O(n^2) Time | O(1) Space
# * n -> The length of nums array


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums_len = len(nums)
        res = []
        if nums_len <= 2:
            return res

        nums.sort()
        for i in range(0, nums_len - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            left, right = i + 1, nums_len - 1
            while left < right:
                cur_sum = nums[i] + nums[left] + nums[right]
                if cur_sum == 0:
                    res.append([nums[i], nums[left], nums[right]])
                    left += 1
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    right -= 1
                    while right > left and nums[right] == nums[right + 1]:
                        right -= 1
                elif cur_sum < 0:
                    left += 1
                else:
                    right -= 1

        return res


# @lc code=end
